java/arcs/core/storage/WriteBack.kt

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+package arcs.core.storage
+
+import androidx.annotation.VisibleForTesting
+import arcs.core.storage.keys.Protocols
+import java.util.concurrent.ExecutorService
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.asCoroutineDispatcher
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+
+/**
+ * A layer to decouple local data updates and underlying storage layers flushes.
+ *
+ * The modern storage stacks including VFS page dirty write-back, io-scheduler
+ * merging and re-ordering, flash device turbo-write and/or buffer flush, etc
+ * are all implemented in an efficient way where no data/metadata is written
+ * through to storage media for every single write operation unless explicitly
+ * being requested.
+ */
+interface WriteBack {
+    /**
+     * Write-through: flush directly all data updates to the next storage layer.
+     */
+    suspend fun flush(job: suspend () -> Unit)
+    /**
+     * Write-back: queue up data updates and let write-back threads decide how and
+     * when to flush all data updates to the next storage layer.
+     */
+    suspend fun asyncFlush(job: suspend () -> Unit)
+
+    /** Await completion of all active flush jobs. */
+    suspend fun awaitIdle()
+}
+
+/** The factory interfaces of [WriteBack] implementations. */
+interface WriteBackFactory {
+    fun create(
+        protocol: String = "",
+        /**
+         * Provide a dedicated write-back thread pool or leave it to implementations
+         * to decide what threads to be designated as the write-backers.
+         */
+        writebackThreads: ExecutorService? = null,
+        /**
+         * The maximum queue size above which new incoming flush jobs will be suspended.
+         */
+        queueSize: Int = 10
+    ): WriteBack
+}
+
+/** Write-back implementation for Arcs Stores.*/
+class StoreWriteBack private constructor(
+    protocol: String,
+    writebackThreads: ExecutorService?,
+    queueSize: Int
+) : WriteBack,
+    Channel<suspend () -> Unit> by Channel(queueSize),
```

Comment:
Might be more readable to just have a local Channel field, instead of making the whole class a Channel.

Reply:
The same reason, let WriteBack provide the Channel interface for outside world to cancel or check status (active, etc).

Reply:
I finally decided to follow your suggestion to make it a local field. Thanks.